File: gosource/indicator/zigzag/zigzag.go
package zigzag

import (
	"global"
	"math"
)

type Zigzag struct {
	Bar_count    int
	Price_list   []global.Price_dic
	Zigzag_list  []global.Cur_obj
	Highlow_list []global.Cur_obj
	Status       int
	cur_high     global.Cur_obj
	cur_low      global.Cur_obj
	inv_max      float64
	inv_min      float64
	pos          int
	prev_pos     int
	preprev_pos  int
}

func (Z *Zigzag) Init() {
	Z.Price_list = make([]global.Price_dic, 0)
	Z.Zigzag_list = make([]global.Cur_obj, 0)
	Z.Highlow_list = make([]global.Cur_obj, 0)
	Z.Status = 0
	Z.cur_high = global.Cur_obj{Date_time: "1970-01-01", Price: math.Inf(-1), Close: math.Inf(-1), Dist: 1, Type: 1}
	Z.cur_low = global.Cur_obj{Date_time: "1970-01-01", Price: math.Inf(1), Close: math.Inf(1), Dist: 1, Type: 0}
	Z.inv_max = Z.cur_high.Price
	Z.inv_min = Z.cur_low.Price
	Z.pos = 0
	Z.prev_pos = 0
	Z.preprev_pos = -1
}

func (Z *Zigzag) Compute(price_dic global.Price_dic) {
	if Z.Status == 1 {
		if price_dic.High > Z.inv_max && price_dic.High > Z.cur_high.Price {
			Z.cur_high.Date_time = price_dic.Date_time
			Z.cur_high.Price = price_dic.High
			Z.cur_high.Close = price_dic.Close
			Z.cur_high.Volumn = price_dic.Volumn
			Z.cur_high.Dist = Z.pos - Z.prev_pos + 1
			Z.prev_pos = Z.pos
			temp_cur := Z.cur_high
			Z.Zigzag_list = append(Z.Zigzag_list, temp_cur)
		} else if price_dic.Low < Z.inv_min {
			Z.cur_low.Date_time = price_dic.Date_time
			Z.cur_low.Price = price_dic.Low
			Z.cur_low.Close = price_dic.Close
			Z.cur_low.Volumn = price_dic.Volumn
			Z.cur_low.Dist = Z.pos - Z.prev_pos + 1
			// 更新当前zigzag的状态为下降
			Z.Status = 0
			Z.prev_pos = Z.pos
			temp_cur := Z.cur_low
			Z.Zigzag_list = append(Z.Zigzag_list, temp_cur)
		}
	} else if Z.Status == 0 {
		if price_dic.Low < Z.inv_min && price_dic.Low < Z.cur_low.Price {
			Z.cur_low.Date_time = price_dic.Date_time
			Z.cur_low.Price = price_dic.Low
			Z.cur_low.Close = price_dic.Close
			Z.cur_low.Volumn = price_dic.Volumn
			Z.cur_low.Dist = Z.pos - Z.prev_pos + 1
			Z.prev_pos = Z.pos
			temp_cur := Z.cur_low
			Z.Zigzag_list = append(Z.Zigzag_list, temp_cur)
		} else if price_dic.High > Z.inv_max {
			Z.cur_high.Date_time = price_dic.Date_time
			Z.cur_high.Price = price_dic.High
			Z.cur_high.Close = price_dic.Close
			Z.cur_high.Volumn = price_dic.Volumn
			Z.cur_high.Dist = Z.pos - Z.prev_pos + 1
			//更新当前状态为上升
			Z.Status = 1
			Z.prev_pos = Z.pos
			temp_cur := Z.cur_high
			Z.Zigzag_list = append(Z.Zigzag_list, temp_cur)
		}
	}
	Z.pos += 1
	// 更新price_list
	if len(Z.Price_list) >= Z.Bar_count {
		Z.Price_list = append(Z.Price_list, price_dic)[1:]
	} else {
		Z.Price_list = append(Z.Price_list, price_dic)
	}
	// 更新区间最大值的最大值和最小值的最小值
	temp_max := math.Inf(-1)
	temp_min := math.Inf(1)
	for i := 0; i < len(Z.Price_list); i++ {
		if Z.Price_list[i].High > temp_max {
			temp_max = Z.Price_list[i].High
		}
		if Z.Price_list[i].Low < temp_min {
			temp_min = Z.Price_list[i].Low
		}
	}
	Z.inv_max = temp_max
	Z.inv_min = temp_min

	if Z.check_hl() {
		if len(Z.Highlow_list) != 0 {
			if Z.Highlow_list[len(Z.Highlow_list)-1] != Z.Highlow_list[len(Z.Highlow_list)-2] {
				Z.Highlow_list = append(Z.Highlow_list, Z.Highlow_list[len(Z.Highlow_list)-2])
			} else {
				Z.Highlow_list = append(Z.Highlow_list, Z.Zigzag_list[len(Z.Zigzag_list)-2])
			}
		}
	}

}

func (Z *Zigzag) check_hl() bool {
	h1_flag := false
	if len(Z.Zigzag_list) > 1 {
		h1_flag = (Z.Zigzag_list[len(Z.Zigzag_list)-1].Type != Z.Zigzag_list[len(Z.Zigzag_list)-2].Type)
	}
	return h1_flag
}

func (Z *Zigzag) Current_zigzag() global.Cur_obj {
	return Z.Zigzag_list[len(Z.Zigzag_list)-1]
}

// 防止zigzag_list 过长
func (Z *Zigzag) Shrunk_size(Max_length int) {
	if len(Z.Zigzag_list) > Max_length {
		Z.Zigzag_list = Z.Zigzag_list[len(Z.Zigzag_list)-Max_length:]
	}
}

File: pysource/Indicators/zigzag.py
#!/usr/bin/env python3
# -*- encoding: utf-8 -*-
# LUX et VERITAS
# Create On: 2022/11/09 09:55:33


from collections import deque


class ZigZag(object):
    """Calculate ZigZag
    
    Usage:
    >>> zzo = ZigZag(20)  # 对于每个品种，首先实例化ZigZag。
    >>> for cur_price in price_data.to_dict(orient="records"):   # 无论是回测还是实盘，都按照时间先后顺序逐个bar进行传入。
    >>>     zzo.compute(cur_price)
    >>>     zigzag_list = zzo.zig_zag_
    >>>     do something......
    """
    
    def __init__(self, bar_count=20, status=0, cur_high: dict={}, cur_low: dict={}, pos=0, prev_pos=0):
        # 计算过去多少根bar的zigzag
        self._bar_count = bar_count
        # 存放bar_count跟bar的价格相关等数据
        self._price_list = deque(maxlen=self._bar_count)
        # zigzag信息列表，基础元素为字典
        self._zigzag_list = list()
        # highlow信息列表，基础元素字典
        self._highlow_list = list()
        # 当前zigzag状态， 0：下降；1：上升
        self._status = status
        # 当前高点价格信息，
        # 样例：{"date_time": date_time, "price": high/low/close, "close": close, "volume": volume, "dist": dist, "type": 1}
        # 0：低点；1：高点
        if len(cur_high) > 0:
            self._cur_high = cur_high.copy()
        else:
            self._cur_high = {
                "date_time": "1970-01-01",
                "price": float("-inf"),
                "close": float("-inf"),
                "volume": None,
                "dist": 1,  # 与前一个高/低点的距离，该距离包括了端点，因此最小为1
                "type": 1,
            }
        # 当前低点价格信息，
        # 样例：{"date_time": date_time, "price": high/low/close, "close": close, "volume": volume, "dist": dist, "type": 0}
        # 0：低点；1：高点
        if len(cur_low) > 0:
            self._cur_low = cur_low.copy()
        else:
            self._cur_low = {
                "date_time": "1970-01-01",
                "price": float("inf"),
                "close": float("inf"),
                "volume": None,
                "dist": 1,  # 与前一个高/低点的距离，该距离包括了端点，因此最小为1
                "type": 0,
            }
        # 初始化区间最大/最小为当前最大/最小
        self._inv_max = self._cur_high["price"]
        self._inv_min = self._cur_low["price"]
        # 当前高低点所在位次
        self._pos = pos
        # 前一个高低点价格位次
        self._prev_pos = prev_pos
        # 初始的位置
        self._preprev_pos = -1

    
    def compute(self, price_dict):
        """Compute
        
        Parameters
        -----------
            price_dict : dictionary
                样例, {
                    "date_time": date_time,
                    "open": open,
                    "high": high,
                    "low": low,
                    "close": close,
                    "volume": volumne
                }
                至少有以上6项
        """

        price_dict = price_dict.copy()
        # zigzag上升状态
        if self._status == 1:
            # 在 当前bar的最高点大于上一个区间最大值且当前价格大于当前高点价格信息时，更新当前高点价格信息
            # 同时将高点价格信息更新到zigzag_list
            if price_dict["high"] > self._inv_max and price_dict["high"] > self._cur_high["price"]:
                self._cur_high["date_time"] = price_dict["date_time"]
                self._cur_high["price"] = price_dict["high"]
                self._cur_high["close"] = price_dict["close"]
                self._cur_high["volume"] = price_dict["volume"]
                self._cur_high["dist"] = self._pos - self._prev_pos + 1  # 这里的距离包括了两个端点
                self._prev_pos = self._pos
                self._zigzag_list.append(self._cur_high.copy())
            # 在 当前bar的最低点小于上一个区间最小值时 更新当前低点价格信息，同时更改zigzag状态为下降
            # 之后，将当前低点价格信息添加到zigzag列表中
            # 注：这里不认为一阴穿Nbar的状态会改变原zigzag状态
            elif price_dict["low"] < self._inv_min:
                self._cur_low["date_time"] = price_dict["date_time"]
                self._cur_low["price"] = price_dict["low"]
                self._cur_low["close"] = price_dict["close"]
                self._cur_low["volume"] = price_dict["volume"]
                self._cur_low["dist"] = self._pos - self._prev_pos + 1  # 这里的距离包括了两个端点
                # 更新当前zigzag的状态为下降
                self._status = 0
                self._prev_pos = self._pos
                self._zigzag_list.append(self._cur_low.copy())
        # zigzag下降状态
        elif self._status == 0:
            # 在 当前bar的最低点小于上一个区间最小值且当前价格小于当前低点价格信息时，更新当前低点价格信息
            # 同时将低点价格信息更新到zigzag_list
            if price_dict["low"] < self._inv_min and price_dict["low"] < self._cur_low["price"]:
                self._cur_low["date_time"] = price_dict["date_time"]
                self._cur_low["price"] = price_dict["low"]
                self._cur_low["close"] = price_dict["close"]
                self._cur_low["volume"] = price_dict["volume"]
                self._cur_low["dist"] = self._pos - self._prev_pos + 1  # 这里的距离包括了两个端点
                self._prev_pos = self._pos
                self._zigzag_list.append(self._cur_low.copy())
            # 在 当前bar的最高点大于上一个区间最大值时 更新当前高点价格信息，同时更改zigzag状态为上升
            # 之后，将当前高点价格信息添加到zigzag列表中
            # 注：这里不认为一阳穿Nbar会改变原zigzag状态
            elif price_dict["high"] > self._inv_max:
                self._cur_high["date_time"] = price_dict["date_time"]
                self._cur_high["price"] = price_dict["high"]
                self._cur_high["close"] = price_dict["close"]
                self._cur_high["volume"] = price_dict["volume"]
                self._cur_high["dist"] = self._pos - self._prev_pos + 1  # 这里的距离包括了两个端点
                # 更新当前zigzag的状态为上升
                self._status = 1
                self._prev_pos = self._pos
                self._zigzag_list.append(self._cur_high.copy())
        
        self._pos += 1
        # 更新区间价格列表，并更新区间最大值和最小值
        self._price_list.append(price_dict)
        self._inv_max = max(self._price_list, key=lambda x: x["high"])["high"]
        self._inv_min = min(self._price_list, key=lambda x: x["low"])["low"]

        # 第一条数据输送进入之后肯定会产生一个zigzag值
        if self._check_hl():
            if len(self._highlow_list):
                if self._highlow_list[-1] != self._zigzag_list[-2]:
                    self._highlow_list.append(self._zigzag_list[-2])
            else:
                self._highlow_list.append(self._zigzag_list[-2])

    def _check_hl(self):
        hl_flag = False
        if len(self._zigzag_list) > 1:
            hl_flag = (self._zigzag_list[-1]["type"] != self._zigzag_list[-2]["type"])

        return hl_flag  
    
    @property
    def zigzag_list_(self):
        return self._zigzag_list

    @property
    def highlow_list_(self):
        return self._highlow_list

    @property
    def current_high_(self):
        return self._cur_high
    
    @property
    def current_low_(self):
        return self._cur_low
    
    @property
    def status_(self):
        return self._status

    @property
    def current_zigzag_(self):
        if len(self._zigzag_list):
            return self._zigzag_list[-1]
            # # Get HL list batch NO USE
            # zz_df = pd.DataFrame(zzo.zigzag_list_)
            # zz_sftdf = zz_df.shift(1, fill_value=0)
            # hl_df = zz_sftdf[(zz_df["type"] != zz_sftdf["type"])]

    @property
    def current_highlow_(self):
        if len(self._highlow_list) > 1:
            return (self._highlow_list[-1], self._highlow_list[-2])

    

